### ISSUE TITLE
Pokémon data structure plans

### ISSUE THREAD

**Issue by Rangi42** [Rangi42/polishedcrystal#219]
(Not starting on this until `optimizations` is merged into `master` and another beta patch has been published.)

- [ ] 20-bit OT ID allows six digits, 000000 to 999999
- [ ] 5-bit IVs replace 4-bit DVs
- [x] IsEgg bit replaces the `EGG` species ID (so Shuckle can return)
- [x] ~~IsDead bit marks death in Nuzlocke mode~~ (obsolete)
- [ ] Move level to party struct, since it derives from experience
- [ ] Move PP to party struct, since it's okay to recover when deposited (not PP Ups)
- [ ] Future-proof space for 9-bit species, item, and move IDs
- [x] Move nickname and OT into the box struct (not shown below)

```diff
 box_struct: MACRO
 \1Species::        db
+\1Personality::
+\1IsEgg::
+\1Shiny::
+\1Form::
+\1SpeciesExtra::   db ; {isEgg:1, shiny:1, form:5, speciesExtra:1}
+\1Gender::
+\1IsDead::
+\1Nature::
+\1ItemExtra::      db ; {gender:1, isDead:1, nature:5, itemExtra:1}
 \1Item::           db
 \1Moves::          ds NUM_MOVES
-\1ID::             dw
+\1ID::
+\1MovesExtra::     ds 3 ; {id:20, movesExtra:4}
 \1Exp::            ds 3
 \1EVs::
 \1HPEV::           db
 \1AtkEV::          db
 \1DefEV::          db
 \1SpdEV::          db
 \1SatEV::          db
 \1SdfEV::          db
-\1DVs::
-\1HPAtkDV::        db
-\1DefSpdDV::       db
-\1SatSdfDV::       db
+\1IVs::            ds 4 ; {hp:5, atk:5, def:5, spd:5, sat:5, sdf:5, unused:2}
+\1Ability::
+\1HyperTraining::  db ; {hp:1, atk:1, def:1, spd:1, sat:1, sdf:1, ability:2}
-\1Personality::
-\1Shiny::
-\1Ability::
-\1Nature::         db
-\1Gender::
-\1IsEgg::
-\1IsDead::
-\1Form::           db
-\1PP::             ds NUM_MOVES
+\1PPUps::          db
 \1Happiness::      db
 \1PokerusStatus::  db
 \1CaughtData::
 \1CaughtGender::
 \1CaughtTime::
 \1CaughtBall::     db ; {gender:1, time:2, ball:5}
 \1CaughtLevel::    db
 \1CaughtLocation:: db
-\1Level::          db
+\1Unused::         db ; keeps the total box_struct size as 32 bytes
 \1End::
 ENDM

 party_struct: MACRO
 	box_struct \1
+\1Level::          db
 \1Status::         db
-\1Unused::         db
+\1PP::             ds 4
 \1HP::             dw
 \1MaxHP::          dw
 \1Stats:: ; big endian
 \1Attack::         dw
 \1Defense::        dw
 \1Speed::          dw
 \1SpclAtk::        dw
 \1SpclDef::        dw
 \1StatsEnd::
 ENDM
```

**Comment by Monstarules** [comment]
If you're going for 9-bit mons, you might as well go for the gold and do full 16-bit. Even changing it to 9-bit would make you have to refactor roughly the same amount of data (according to ax6 and others).

**Comment by FredrIQ** [comment]
If we're refactoring the data structure in the future, we ought to make nickname and OT part of the struct instead of seperately, which is a bit silly.

There's also the redundant species byte, but that is a bit more understandable since it makes menu functions happy, but it would be nice to get rid of that as well.

**Comment by Rangi42** [comment]
True; although moving the nickname and OT would be about neatness, not increased capabilities, so it's not a priority.

Getting rid of the redundant six-species byte array would also be nice, especially since getting rid of Eggs is part of that.

**Comment by CiriousJoker** [comment]
Are there memory constraints to this? If there aren't, why not just keep as much info as possible for future proofing?

**Comment by CiriousJoker** [comment]
@Rangi42

Has this thing been implemented? I'm asking because this woulf likely cause major incompatibility problems.

**Comment by FredrIQ** [comment]
It might be worth moving capture metadata to the 3 unused bytes for OT name. They're both used in basically the same circumstances, and due to how infrequently it's used, would probably require the least amount of work. This frees up 3 additional bytes in the box struct.

**Comment by FredrIQ** [comment]
Related: This is an idea for how to make the Storage System experience smoother.

* Pokémon in the Storage System currently use 20 bytes, 11*2 with nickname+OT, 40 bytes for all of it with string terminators and redundant species byte disregarded.
* Store each Pokémon's data only to which it is needed in the 2 save banks. Disregarding limitations caused by storing them as entire Boxes, you can fit 157 Pokémon. With a bit of space for metadata, this can be limited to 150 per bank.
* Alongside the Pokémon themselves, store a bit array in each bank for which of the 150 entries are allocated to store Pokémon, and which are just free space
* Repurpose the Active Box (1348 bytes) section in the save to instead store data about which Pokémon entries each of the 15 (changed from the previous 14) Boxes is referencing. This takes 600 bytes (2 bytes per box entry).
* For the rest of the unused bytes in Active Box, allocate 10 bytes for an ongoing Storage System operation (moving stuff around, etc).
  - 1 byte to denote that we're doing something, and how far into the process we are
  - up to 9 bytes (probably less, but just in case) describing the change we are applying
* Allocate the rest of the bytes to a journal that describes *in reverse* the actions we've done to the Storage System overall since our last save
  - This allows us to, upon restarting the game without saving, applying those steps to rollback the changes made to the Storage System to the point it was in the last save.
* Only when this space runs out (very unlikely), force the player to save the game to continue using the Storage System If they refuse, disallow capturing Pokémon with a full party until the next time they save.
* When saving the game, once we've saved the game data, all we need to do is to tell the game that the current storage system state is final, and to flush the activity log.
